File: src/main/java/io/tatum/ledger/Account.java
package io.tatum.ledger;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.base.Strings;
import io.tatum.model.request.BlockAmount;
import io.tatum.model.request.CreateAccount;
import io.tatum.model.request.CreateAccountsBatch;
import io.tatum.model.response.ledger.AccountBalance;
import io.tatum.model.response.ledger.Blockage;
import io.tatum.model.response.ledger.BookkeepingAccount;
import io.tatum.utils.Async;
import io.tatum.utils.Env;

import java.io.IOException;
import java.util.concurrent.ExecutionException;

import static io.tatum.constants.Constant.TATUM_API_URL;

public class Account {

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/getAccountByAccountId" target="_blank">Tatum API documentation</a>
     */
    public BookkeepingAccount getAccountById(String id) throws IOException, ExecutionException, InterruptedException {
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account" + id;
        String account = Async.get(uri, Env.getTatumApiKey());
        // TO-DO
        return new BookkeepingAccount();
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/createAccount" target="_blank">Tatum API documentation</a>
     */
    public BookkeepingAccount createAccount(CreateAccount account) throws IOException, ExecutionException, InterruptedException {
//        await validateOrReject(account);
//        TO-DO
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account";

        var objectMapper = new ObjectMapper();
        String requestBody = objectMapper.writeValueAsString(account);

        String acc = Async.post(uri, Env.getTatumApiKey(), requestBody);
        // TO-DO
        return new BookkeepingAccount();
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/createAccountBatch" target="_blank">Tatum API documentation</a>
     */
    public BookkeepingAccount[] createAccounts(CreateAccountsBatch accounts) throws IOException, ExecutionException, InterruptedException {
//        await validateOrReject(accounts);
//        TO-DO
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account/batch";
        var objectMapper = new ObjectMapper();
        String requestBody = objectMapper.writeValueAsString(accounts);

        String acc = Async.post(uri, Env.getTatumApiKey(), requestBody);
        // TO-DO
        return new BookkeepingAccount[]{};
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/getBlockAmount" target="_blank">Tatum API documentation</a>
     */
    public Blockage[] getBlockedAmountsByAccountId(String id, Integer pageSize, Integer offset) throws IOException, ExecutionException, InterruptedException {
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account/block/" + id + "?pageSize=" + pageSize + "&offset=" + offset;
        String tx = Async.get(uri, Env.getTatumApiKey());
        // TO-DO
        return new Blockage[]{};
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/blockAmount" target="_blank">Tatum API documentation</a>
     */
    public String blockAmount(String id, BlockAmount block) throws IOException, ExecutionException, InterruptedException {
//        await validateOrReject(block);
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account/block/" + id;
        var objectMapper = new ObjectMapper();
        String requestBody = objectMapper.writeValueAsString(block);

        String blockId = Async.post(uri, Env.getTatumApiKey(), requestBody);
        return blockId;
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/deleteBlockAmount" target="_blank">Tatum API documentation</a>
     */
    public void deleteBlockedAmount(String id) throws IOException, ExecutionException, InterruptedException {
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account/block/" + id;
        Async.delete(uri, Env.getTatumApiKey());
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/deleteAllBlockAmount" target="_blank">Tatum API documentation</a>
     */
    public void deleteBlockedAmountForAccount(String id) throws IOException, ExecutionException, InterruptedException {
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account/block/acoount/" + id;
        Async.delete(uri, Env.getTatumApiKey());
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/activateAccount" target="_blank">Tatum API documentation</a>
     */
    public void activateAccount(String id) throws IOException, ExecutionException, InterruptedException {
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account/" + id + "/activate";
        String requestBody = "{}";
        Async.put(uri, Env.getTatumApiKey(), requestBody);
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/deactivateAccount" target="_blank">Tatum API documentation</a>
     */
    public void deactivateAccount(String id) throws IOException, ExecutionException, InterruptedException {
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account/" + id + "/deactivate";
        String requestBody = "{}";
        Async.put(uri, Env.getTatumApiKey(), requestBody);
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/freezeAccount" target="_blank">Tatum API documentation</a>
     */
    public void freezeAccount(String id) throws IOException, ExecutionException, InterruptedException {
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account/" + id + "/freeze";
        String requestBody = "{}";
        Async.put(uri, Env.getTatumApiKey(), requestBody);
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/unfreezeAccount" target="_blank">Tatum API documentation</a>
     */
    public void unfreezeAccount(String id) throws IOException, ExecutionException, InterruptedException {
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account/" + id + "/unfreeze";
        String requestBody = "{}";
        Async.put(uri, Env.getTatumApiKey(), requestBody);
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/getAccountsByCustomerId" target="_blank">Tatum API documentation</a>
     */
    public BookkeepingAccount[] getAccountsByCustomerId(String id, Integer pageSize, Integer offset) throws IOException, ExecutionException, InterruptedException {
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account/customer/" + id + "?pageSize=" + pageSize + "&offset=" + offset;
        String accounts = Async.get(uri, Env.getTatumApiKey());
        // TO-DO
        return new BookkeepingAccount[]{};
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/getAllAccounts" target="_blank">Tatum API documentation</a>
     */
    public BookkeepingAccount[] getAllAccounts(Integer pageSize, Integer offset) throws IOException, ExecutionException, InterruptedException {
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account?pageSize=" + pageSize + "&offset=" + offset;
        String accounts = Async.get(uri, Env.getTatumApiKey());
        // TO-DO
        return new BookkeepingAccount[]{};
    }

    /**
     * For more details, see <a href="https://tatum.io/apidoc#operation/getAccountBalance" target="_blank">Tatum API documentation</a>
     */
    public AccountBalance getAccountBalance(String id) throws IOException, ExecutionException, InterruptedException {
        String tatumApiUrl = Env.getTatumApiUrl();
        String uri = (Strings.isNullOrEmpty(tatumApiUrl) ? TATUM_API_URL : tatumApiUrl) + "/v3/ledger/account/" + id + "/balance";
        String accounts = Async.get(uri, Env.getTatumApiKey());
        // TO-DO
        return new AccountBalance();
    }

}

File: src/main/kotlin/ledger/account.kt
package ledger

import deleteFrom
import getArrayFrom
import getObjectFrom
import model.request.BlockAmount
import model.request.CreateAccount
import model.response.ledger.Account
import model.response.ledger.AccountBalance
import model.response.ledger.Blockage
import model.response.ledger.IdString
import postObjectTo
import putTo

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/getAccountByAccountId">Tatum API documentation</a>
 */
fun getAccountById(accountId: String): Account = getObjectFrom("/ledger/account/$accountId")

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/createAccount">Tatum API documentation</a>
 */
fun createAccount(account: CreateAccount): Account = postObjectTo(
    endpoint = "/ledger/account",
    payload = account,
)

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/getBlockAmount">Tatum API documentation</a>
 */
fun getBlockedAmountsByAccountId(accountId: String, pageSize: Long = 50, offset: Long = 0): Array<Blockage> =
    getArrayFrom(
        endpoint = "/ledger/account/block/$accountId",
        params = mapOf("pageSize" to pageSize.toString(), "offset" to offset.toString()),
    )

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/blockAmount">Tatum API documentation</a>
 */
fun blockAmount(accountId: String, block: BlockAmount): IdString =
    postObjectTo(
        endpoint = "/ledger/account/block/$accountId",
        payload = block
    )

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/deleteBlockAmount">Tatum API documentation</a>
 */
fun deleteBlockedAmount(blockageId: String) = deleteFrom("/ledger/account/block/$blockageId")

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/deleteAllBlockAmount">Tatum API documentation</a>
 */
fun deleteBlockedAmountForAccount(accountId: String) = deleteFrom("/ledger/account/block/account/$accountId")

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/activateAccount">Tatum API documentation</a>
 */
fun activateAccount(accountId: String) = putTo("/ledger/account/$accountId/activate")

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/deactivateAccount">Tatum API documentation</a>
 */
fun deactivateAccount(accountId: String) = putTo("/ledger/account/$accountId/deactivate")

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/freezeAccount">Tatum API documentation</a>
 */
fun freezeAccount(accountId: String) = putTo("/ledger/account/$accountId/freeze")

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/unfreezeAccount">Tatum API documentation</a>
 */
fun unfreezeAccount(accountId: String) = putTo("/ledger/account/$accountId/unfreeze")

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/getAccountsByCustomerId">Tatum API documentation</a>
 */
fun getAccountsByCustomerId(customerId: String, pageSize: Long = 50, offset: Long = 0): Array<Account> =
    getArrayFrom(
        endpoint = "/ledger/account/customer/$customerId",
        params = mapOf("pageSize" to pageSize.toString(), "offset" to offset.toString()),
    )

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/getAllAccounts">Tatum API documentation</a>
 */
fun getAllAccounts(pageSize: Long = 50, offset: Long = 0): Array<Account> =
    getArrayFrom(
        endpoint = "/ledger/account",
        params = mapOf("pageSize" to pageSize.toString(), "offset" to offset.toString()),
    )

/**
 * For more details see link
 * @see <a href="https://tatum.io/apidoc.html#operation/getAccountBalance">Tatum API documentation</a>
 */
fun getAccountBalance(accountId: String): AccountBalance = getObjectFrom("/ledger/account/$accountId/balance")
